numba/np/npyimpl.py

```diff
@@ -297,7 +302,8 @@ def raise_impl(self_shape, other_shape):
         )
         tup = (context.make_tuple(builder, sig.args[0], self.shape),
                context.make_tuple(builder, sig.args[1], other.shape),)
-        context.compile_internal(builder, raise_impl, sig, tup)
+        context.typing_context.refresh()
```

Comment:
Why does the typing context need refreshing?

Reply:
Function is declared inside the scope of `guard_match_core_dim`. If the typing context is not refreshed, Numba fails at a later stage:

```
  File "/Users/guilhermeleobas/git/numba/numba/np/npyimpl.py", line 624, in numpy_gufunc_kernel
    arg_a.guard_match_core_dims(ufunc, arg_b, idx_a, idx_b)
  File "/Users/guilhermeleobas/git/numba/numba/np/npyimpl.py", line 308, in guard_match_core_dims
    context.call_overload(builder, raise_impl, sig, tup)
  File "/Users/guilhermeleobas/git/numba/numba/core/base.py", line 911, in call_overload
    fnty = self.typing_context.resolve_value_type(impl)
  File "/Users/guilhermeleobas/git/numba/numba/core/typing/context.py", line 397, in resolve_value_type
    raise typeof_exc
ValueError: Cannot determine Numba type of <class 'function'>
```

Reply:
OK. If it's possible for the `call_overload` function to fail due to inconsistent state in the typing registries, should a typing context refresh be made unconditionally at the start of the the `call_overload` function implementation to prevent this from ever happening?

Reply:
Wouldn't that be too slow? Refreshing the typing context is not needed if the function is declared in the global scope.